src/aggregator/client/queue.go

```diff
@@ -208,24 +209,63 @@ func (q *queue) Close() error {
 }
 
 func (q *queue) Flush() {
+	var (
+		buf = _queueConnWriteBufPool.Get().(*[]byte)
+		n   int
+		err error
+	)
+
+	for err == nil {
+		// flush everything in batches, to make sure no single payload is too large,
+		// to prevent a) allocs and b) timeouts due to big buffer IO taking too long.
+		var processed int
+		processed, err = q.flush(buf)
+		n += processed
+	}
+
+	if err != nil && err != io.EOF {
+		q.log.Error("error writing data",
+			zap.String("target_instance_id", q.instance.ID()),
+			zap.String("target_instance", q.instance.Endpoint()),
+			zap.Int("bytes_processed", n),
+			zap.Error(err),
+		)
+	}
+
+	// Check buffer capacity, not length, to make sure we're not pooling slices that are too large.
+	// Otherwise, it could result in multi-megabyte slices hanging around, in case we get a single massive write.
+	if cap(*buf) < _queueMaxWriteBufSize {
+		*buf = (*buf)[:0]
+		_queueConnWriteBufPool.Put(buf)
+	}
+}
+
+func (q *queue) flush(buf *[]byte) (int, error) {
+	var n int
+
 	q.mtx.Lock()
 
 	if q.buf.size() == 0 {
 		q.mtx.Unlock()
-		return
+		return n, io.EOF
 	}
 
-	buf := _queueConnWriteBufPool.Get().(*[]byte)
-
+	*buf = (*buf)[:0]
 	for q.buf.size() > 0 {
-		b := q.buf.shift()
+		b := q.buf.peek()
+		if n > 0 && len(b.Bytes())+len(*buf) >= _queueMaxWriteBufSize {
+			// only merge buffers that are smaller than _queueMaxWriteBufSize bytes
+			break
+		}
+		_ = q.buf.shift()
 
 		bytes := b.Bytes()
 		if len(bytes) == 0 {
```

Comment:
nit: this check can be earlier

Reply:
it can't - otherwise we'd have an infinite loop on empty bytes buffer (although it should not happen...)